### PR TITLE
dsv4(npu): support prefill context parallelism with interleave(with layersplit) and zigzag

### PR DESCRIPTION
**Pull request by vstone-w**
## Motivation

1、support prefill cp for deepseekv4 on npu platform
2、support layersplit(cp-interleave) for deepseekv4 on npu platform

## Modifications

1. Based on the latest community CP code, adapt PCP for DeepSeek-V4:

- Enable the CP switch on the NPU platform.

- Perform metadata management inside the DeepSeek-V4 model files, managing two sets—global and local rank—to adapt to the inputs of different operators.

2. Enable the LayerSplit feature of DeepSeek-V4 based on the CP code:
- Add a new KV cache pool for LayerSplit and a new length retrieval interface to adapt to LayerSplit's KV cache sharding characteristics.
 - Use layerid to correspond to different KV cache clusters, and match by layerid during PD transmission.
 - Add LayerSplit plan and cache class encapsulation to provide sharding query functionality.

## Accuracy Tests

image

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #34762289116](https://github.com/sgl-project/sglang/actions/runs/34762289116)
Latest PR Test (Extra): :x: [Run #34762288993](https://github.com/sgl-project/sglang/actions/runs/34762288993)
Latest PR Test (AMD ROCm 10): :x: [Run #34762289112](https://github.com/sgl-project/sglang/actions/runs/34762289112)

### CODE REVIEW

File: python/sglang/srt/disaggregation/prefill.py [hunk 1]
```diff
@@ -241,11 +241,18 @@ def _init_kv_manager(self) -> CommonKVManager:
...
-            else getattr(self.token_to_kv_pool, "end_layer", None)
-        )
+        if layer_shard_enabled:
+            # Bucket-layout pools (e.g. DSV4) expose the owned global layer
+            # range directly; len() covers one-buffer-per-layer pools.
+            kv_args.prefill_end_layer = getattr(
+                self.token_to_kv_pool,
+                "layer_shard_end",
+                kv_args.prefill_start_layer + len(kv_data_ptrs),
+            )
```
Comment (AndyLi429) [inline python/sglang/srt/disaggregation/prefill.py:251]: necessary?

Comment (vstone-w) [reply python/sglang/srt/disaggregation/prefill.py:251]: yes，dsv4 use bucket to manage kvpool，meaning len(kv_data_ptrs) = layer * 3【c4, c4 index, c4_scale】,which is not right

**Comment by AndyLi429** [comment]
how to launch cp and layer split feature.Please attach launch command in your description

**Comment by AndyLi429** [comment]
@codex review

File: python/sglang/srt/arg_groups/parallel_hook.py [hunk 2]
```diff
@@ -618,12 +618,10 @@ def handle_expert_distribution_metrics(server_args: Any):
...
 
 def validate_prefill_cp_platform(server_args: Any):
-    """Reject deprecated platform CP before resolving models or CP topology."""
+    """Reject deprecated platform CP; NPU prefill CP stays supported."""
     cfg = resolving_view(server_args)
     platform = get_platform()
-    if cfg.enable_prefill_cp and (
-        platform.is_hip or platform.is_npu or platform.is_musa
-    ):
+    if cfg.enable_prefill_cp and (platform.is_hip or platform.is_musa):
```
Comment (ClownBin) [inline python/sglang/srt/arg_groups/parallel_hook.py:624]: [P1] Keep the generic CP completion path device-neutral

This change makes the NPU route reach `EagerRunner._execute_extend_cp`, which still unconditionally evaluates `torch.cuda.current_stream()` immediately before `cp_gather_after_forward` (`eager_runner.py:418`). On Ascend-only installs, that fails after the model body and before logits. The CP gather implementations do not use this stream argument; remove it or get the stream from the active device.

File: python/sglang/srt/layers/cp/interleave.py [hunk 3]
```diff
@@ -205,6 +213,18 @@ def _gather_interleaved_tensor(self, x: Any, forward_batch) -> Any:
...
+        actual = metadata.per_rank_actual_token
+        if total_tokens == self.cp_size * physical_rank_len and all(
+            int(n) == physical_rank_len for n in actual
+        ):
+            return (
+                gathered.view(self.cp_size, physical_rank_len, *x.shape[1:])
+                .transpose(0, 1)
+                .reshape(total_tokens, *x.shape[1:])
+            )
+
```
Comment (vstone-w) [inline python/sglang/srt/layers/cp/interleave.py:227]: print shape，check this

File: python/sglang/srt/models/deepseek_v4.py [hunk 4]
```diff
@@ -3059,6 +3093,7 @@ def _forward_layers_tbo(
         )
         return hidden_states
 
+    @torch.no_grad()
```
Comment (vstone-w) [inline python/sglang/srt/models/deepseek_v4.py:3096]: delete it

File: python/sglang/srt/models/deepseek_v4.py [hunk 5]
```diff
@@ -3100,9 +3135,16 @@ def forward(
...
+
+        attn_backend = get_attn_backend()
+        if _is_npu and forward_batch.attn_cp_metadata is not None:
+            attn_backend.prepare_dsv4_cp_metadata(forward_batch)
+            local_positions = getattr(forward_batch, "dsv4_cp_local_positions", None)
+            if (
+                local_positions is not None
+                and positions.shape[0] == local_positions.shape[0]
+            ):
+                forward_batch.positions = positions
```
Comment (vstone-w) [inline python/sglang/srt/models/deepseek_v4.py:3147]: 1. gpu metadata managemant
2. fix comment

File: python/sglang/srt/models/deepseek_v4.py
Comment (vstone-w) [inline python/sglang/srt/models/deepseek_v4.py:1545]: check cp v2

File: python/sglang/srt/disaggregation/utils.py [hunk 6]
```diff
@@ -1439,6 +1445,43 @@ def append_dsa_tail(pool) -> None:
                         c128_ptrs,
                         c128_lens,
                         c128_item_lens,
+                        layer_ids=(
+                            token_to_kv_pool.get_c128_layer_ids()
+                            if hasattr(token_to_kv_pool, "get_c128_layer_ids")
+                            else None
+                        ),
+                    )
+            # NPU c128 KV buffers ride as their own ascend-only component.
```
Comment (vstone-w) [inline python/sglang/srt/disaggregation/utils.py:1454]: use isinstance instead of hasattr
